### DISCUSSION TITLE
Listen to gRPC UNARY_STREAM Connection

### DISCUSSION THREAD

**Discussion by gandhishripala** [grpc-requests/grpc_requests#74] [General]
I want to make connection to gRPC UNARY_STREAM and then list to response of Stream,

Example scenario:
1. Open connection to greeter_service.SayHelloGroup
2. Call other methods
3. Read response from greeter_service.SayHelloGroup
4. Repeat steps from 2-3

```
    client = AsyncClient(
        "localhost:50051", descriptor_pool=descriptor_pool.DescriptorPool()
    )
    greeter_service = await client.service("helloworld.Greeter")
    async with greeter_service.SayHelloGroup({"name": "sinsky"}) as greeter_connection:
        response = await anext(greeter_connection)
        assert response
```
Can someone guide me on how this can be achieved?

**Comment by aidandj** [comment]
Does this work?

    client = AsyncClient(
        "localhost:50051", descriptor_pool=descriptor_pool.DescriptorPool()
    )
    greeter_service = await client.service("helloworld.Greeter")
    async for response in greeter_service.SayHelloGroup({"name": "sinsky"}):
        assert response

**Reply by gandhishripala** [reply]
Thank you, I tried this, getting error: ```TypeError: 'async for' requires an object with __aiter__ method, got coroutine```. Any pointer to resolve this?

**Reply by aidandj** [reply]
Share your code. That sounds like you aren't actually invoking the coroutine, are you missing parenthesis and arguments?

**Reply by aidandj** [reply]
Oh sorry, I forgot something. You need to await the coroutine

```
client = AsyncClient(
    "localhost:50051", descriptor_pool=descriptor_pool.DescriptorPool()
)
greeter_service = await client.service("helloworld.Greeter")
async for response in await greeter_service.SayHelloGroup({"name": "sinsky"}):
    assert response
```

**Reply by gandhishripala** [reply]
Sorry for delayed response, this worked. Thank you, I missed await in for statement.

**Reply by gandhishripala** [reply]
Another query, how do we close this connection? I tried with break in async for loop, it throws message (don't have exact message can share that) that thread is waiting for response and connection is closed.

**Reply by gandhishripala** [reply]
Exact Message is: Task was destroyed but it is pending!

**Reply by wynquinox** [reply]
Apologies for the long delay on noticing the discussion @gandhishripala .
I believe that the client will close on its own at the end of your routine, provided you have waited for, and received, all responses expected.

If you're in a situation where the streaming side of the connection will stream until hung up upon, you may want to explore setting a client max connection time in the client config, or setting some other criterion your loop checks upon which where you force a closure of the channel or client.

As for `Task was destroyed but it is pending` - I believe that error means that the coroutine didn't have time to finish all its tasks before the loop was broken.

**Reply by gandhishripala** [reply]
Thank you for your inputs, @ViridianForge . Also Thank you @artificial-aidan .
